src/test/regress/sql/multi_modifying_xacts.sql

```diff
@@ -0,0 +1,57 @@
+
+ALTER SEQUENCE pg_catalog.pg_dist_shardid_seq RESTART 1200000;
+ALTER SEQUENCE pg_catalog.pg_dist_jobid_seq RESTART 1200000;
+
+
+-- ===================================================================
+-- test end-to-end modification functionality
+-- ===================================================================
+
+CREATE TABLE researchers (
+	id bigint NOT NULL,
+    lab_id int NOT NULL,
+	name text NOT NULL
+);
+
+SELECT master_create_distributed_table('researchers', 'lab_id', 'hash');
+SELECT master_create_worker_shards('researchers', 2, 2);
+
+-- add some data
+INSERT INTO researchers VALUES (1, 1, 'Donald Knuth');
+INSERT INTO researchers VALUES (2, 1, 'Niklaus Wirth');
+INSERT INTO researchers VALUES (3, 2, 'Tony Hoare');
+INSERT INTO researchers VALUES (4, 2, 'Kenneth Iverson');
+
+-- replace a researcher, reusing their id
+BEGIN;
+DELETE FROM researchers WHERE lab_id = 1 AND id = 2;
+INSERT INTO researchers VALUES (2, 1, 'John Backus');
+COMMIT;
+
+SELECT name FROM researchers WHERE lab_id = 1 AND id = 2;
+
+-- abort a modification
+BEGIN;
+DELETE FROM researchers WHERE lab_id = 1 AND id = 1;
+ABORT;
+
+SELECT name FROM researchers WHERE lab_id = 1 AND id = 1;
+
+-- creating savepoints should work...
+BEGIN;
+INSERT INTO researchers VALUES (5, 3, 'Dennis Ritchie');
+SAVEPOINT hire_thompson;
+INSERT INTO researchers VALUES (6, 3, 'Ken Thompson');
+COMMIT;
+
+SELECT name FROM researchers WHERE lab_id = 3 AND id = 6;
+
+-- but rollback should not
+BEGIN;
+INSERT INTO researchers VALUES (7, 4, 'Jim Gray');
+SAVEPOINT hire_engelbart;
+INSERT INTO researchers VALUES (8, 4, 'Douglas Engelbart');
+ROLLBACK TO hire_engelbart;
+COMMIT;
+
+SELECT name FROM researchers WHERE lab_id = 4;
```

Comment:
Completely agreed. Will have an update soon. The "start a transaction and then do something on the worker that causes it to fail" is a little tougher here since (a) changing the connection endpoint during a test will kill the transaction, and (b) even if I could multiplex in some way, e.g. dropping the table and other operations could result in blocking (due to locks and the ongoing open transaction).

I'll push up some test before tomorrow and I'll ping you when they're up. Let me know if they're sufficient.